src/zenml/integrations/neptune/experiment_trackers/run_state.py

```diff
@@ -0,0 +1,106 @@
+import functools
+import hashlib
+from typing import List
+
+import neptune.new as neptune
+
+from zenml.client import Client
+from zenml.integrations.constants import NEPTUNE
+from zenml.steps.base_step import BaseStepMeta
+
+
+def singleton(class_):
+    instances = {}
+
+    def getinstance(*args, **kwargs):
+        if class_ not in instances:
+            instances[class_] = class_(*args, **kwargs)
+        return instances[class_]
+
+    return getinstance
+
+
+class NoActiveRunException(Exception):
+    pass
+
+
+class InvalidExperimentTrackerSelected(Exception):
+    pass
+
+
+@singleton
+class RunProvider:
```

Comment:
We already have some utils for singletons in ZenML, I think it makes sense to use those here as well:
```python
from zenml.utils.singleton import SingletonMetaClass

class RunProvider(metaclass=SingletonMetaClass):
    ...
```

Reply:
Nice! Will do :)